src/mcnp/mcnp_funcs.cpp

```diff
@@ -0,0 +1,805 @@
+
+
+#ifdef TRACE_DAGMC_CALLS
+  // compute and report the angle between old and new
+  moab::CartVect oldv(last_uvw);
+  moab::CartVect newv(*uuu, *vvv, *www);
+
+  std::cout << "surf_reflection: " << angle(oldv, newv)*(180.0 / M_PI) << std::endl;;
+#endif
+
+  // a surface was visited
+  visited_surface = true;
+
+  bool update = true;
+  if (*verify_dir_change) {
+    if (last_uvw[0] == *uuu && last_uvw[1] == *vvv && last_uvw[2] == *www)
+      update = false;
+  }
+
+  if (update) {
+    last_uvw[0] = *uuu;
+    last_uvw[1] = *vvv;
+    last_uvw[2] = *www;
+    history.reset_to_last_intersection();
+  }
+
+#ifdef TRACE_DAGMC_CALLS
+  else {
+    // mark it in the log if nothing happened
+    std::cout << "(noop)";
+  }
+
+  std::cout << std::endl;
+#endif
+
+}
+
+void dagmc_particle_terminate_() {
+  history.reset();
+
+#ifdef TRACE_DAGMC_CALLS
+  std::cout << "particle_terminate:" << std::endl;
+#endif
+}
+
+// *ih              - volue index
+// *uuu, *vvv, *www - ray direction
+// *xxx, *yyy, *zzz - ray point
+// *huge            - passed to ray_fire as 'huge'
+// *dls             - output from ray_fire as 'dist_traveled'
+// *jap             - intersected surface index, or zero if none
+// *jsu             - previous surface index
+void dagmctrack_(int* ih, double* uuu, double* vvv, double* www, double* xxx,
+                 double* yyy, double* zzz, double* huge, double* dls, int* jap, int* jsu,
+                 int* nps) {
+  // Get data from IDs
+  moab::EntityHandle vol = DAG->entity_by_index(3, *ih);
+  moab::EntityHandle prev = DAG->entity_by_index(2, *jsu);
+  moab::EntityHandle next_surf = 0;
+  double next_surf_dist;
+
+#ifdef ENABLE_RAYSTAT_DUMPS
+  moab::OrientedBoxTreeTool::TrvStats trv;
+#endif
+
+  double point[3] = {*xxx, *yyy, *zzz};
+  double dir[3]   = {*uuu, *vvv, *www};
+
+  /* detect streaming or reflecting situations */
+  if (last_nps != *nps || prev == 0) {
+    // not streaming or reflecting: reset history
+    history.reset();
+#ifdef TRACE_DAGMC_CALLS
+    std::cout << "track: new history" << std::endl;
+#endif
+
+  } else if (last_uvw[0] == *uuu && last_uvw[1] == *vvv && last_uvw[2] == *www) {
+    // streaming -- use history without change
+    // unless a surface was not visited
+    if (!visited_surface) {
+      history.rollback_last_intersection();
+#ifdef TRACE_DAGMC_CALLS
+      std::cout << "     : (rbl)" << std::endl;
+#endif
+    }
+#ifdef TRACE_DAGMC_CALLS
+    std::cout << "track: streaming " << history.size() << std::endl;
+#endif
+  } else {
+    // not streaming or reflecting
+    history.reset();
+
+#ifdef TRACE_DAGMC_CALLS
+    std::cout << "track: reset" << std::endl;
+#endif
+
+  }
+
+  moab::ErrorCode result = DAG->ray_fire(vol, point, dir,
+                                         next_surf, next_surf_dist, &history,
+                                         (use_dist_limit ? dist_limit : 0)
+#ifdef ENABLE_RAYSTAT_DUMPS
```

Comment:
Readibility: Probably a bit cleaner to put a variable inside the `#ifdef` with a value, then put that variable here as a parameter, rather than break up the function parameters with a `#ifdef`.